rule/package_directory_mismatch.go

```diff
@@ -0,0 +1,51 @@
+package rule
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// PackageDirectoryMismatchRule detects when package name doesn't match directory name.
+type PackageDirectoryMismatchRule struct{}
+
+var skipDirs = map[string]struct{}{
+	".":        {},
+	"/":        {},
+	"":         {},
+	"internal": {},
+	"testdata": {},
+}
+
+// Apply applies the rule to the given file.
+func (*PackageDirectoryMismatchRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	if file.IsTest() || file.Pkg.IsMain() {
+		return nil
+	}
+
+	dirPath := filepath.Dir(file.Name)
+	dirName := filepath.Base(dirPath)
+	if _, found := skipDirs[dirName]; found {
+		return nil
+	}
```

Comment:
I'm unsure this would work for a file like

testdata/noregression/foo.go

```go
package foo
```

I mean this should be ignored 👆 
It's common that testdata are full of random file, especially in Go linter code.

Please note, that on the contrary this should be detected 👇

internal/foo/something.go

```go
package whatever // MATCH
```

And this shouldn't 👇

internal/bar/quz.go
```go
package bar
```

Reply:
Please note, I'm not even sure whether this is valid, nor if it should be reported

internal/abcd.go
```go
package bar
```

I have no computer with me while I'm doing the review. Please check if this file can be imported and what this package should be

Any idea @alexandear ?

Reply:
> I'm unsure this would work for a file like
> 
> testdata/noregression/foo.go
> 
> ```go
> package foo
> ```
> 
> I mean this should be ignored 👆 It's common that testdata are full of random file, especially in Go linter code.

Agree. Let's ignore any matches inside `testdata`.